wallet/src/repository/wallet/mod.rs

```diff
@@ -327,52 +332,103 @@ where
 
         // Total amount of state and db transactions
         let total = state.transaction_next_id + u32::try_from(state.local_movements.len()).unwrap();
-        let mut transactions: Vec<model::BalanceMovement> = Vec::with_capacity(total as usize);
-
-        // Prepend local pending balance movements (not yet included in blocks)
-        let local_movements = state.local_movements.values().cloned().collect::<Vec<_>>();
-        transactions.extend_from_slice(&local_movements);
-
-        // Prepend balance movements of pending blocks
-        state.pending_movements.values().for_each(|x| {
-            transactions.extend_from_slice(x);
-        });
+        let mut keyed_transactions: Vec<model::BalanceMovement> = Vec::new();
 
         // Query database `transaction_next_id` to compute total amount of transactions
         let db_total = self
             .db
             .get_or_default::<_, u32>(&keys::transaction_next_id(account))?;
 
-        if db_total > 0 {
-            let end = db_total.saturating_sub(offset);
-            let start = end.saturating_sub(limit);
-            let range = start..end;
+        // get number of non-repated pending movements.
+        let pending_length = {
+            let mut glob_acc: usize = 0;
+            state.pending_movements.values().for_each(|x| {
+                glob_acc = glob_acc.saturating_add(x.len());
+            });
+            glob_acc
+        };
+        let total_local = state.local_movements.len();
+
+        // Lets get the ranges for pending and db transactions
+        let (range_local, range_pending, range_db) = calculate_transaction_ranges(
+            offset as usize,
+            limit as usize,
+            total_local,
+            pending_length,
+            db_total as usize,
+        );
 
-            log::debug!(
-                "Retrieving database transactions in range {:?}. Start({}), End({}), Total({})",
-                range,
-                start,
-                end,
-                db_total,
-            );
-            for index in range.rev() {
-                match self.get_transaction(account, index) {
-                    Ok(transaction) => {
-                        transactions.push(transaction);
-                    }
-                    Err(e) => {
-                        log::error!(
-                            "Error while retrieving transaction with index {}: {}",
-                            index,
-                            e
-                        );
+        // Append local movements if any
+        if let Some(range_local) = range_local {
+            // Append local pending balance movements (not yet included in blocks)
+            let mut local_movements: Vec<model::BalanceMovement> =
+                state.local_movements.values().cloned().collect();
+            local_movements.sort_by(|a, b| a.db_key.cmp(&b.db_key));
+            keyed_transactions.extend_from_slice(&local_movements[range_local]);
```

Comment:
In this block of code we are cloning the `model::BalanceMovement` two times: in `.cloned()` and also in `extend_from_slice`. Not sure what's the best way to avoid that, but the simplest one to use `drain` to move the elements from one vector into the other one:

```suggestion
            keyed_transactions.extend(local_movements.drain(range_local));
```